File: trees1/lowestcommonancestor.cpp
#include<bits/stdc++.h>
using namespace std;
class node
{
    public:
        int data;
        node* left;
        node* right;
        node(int data1)
        {
            data=data1;
            left=right=nullptr;
        }
};  
node* lowestcommonancestor(node* root,node* p,node* q)
{
    if(root==NULL||root==p||root==q)
    {
        return root;
    }
    node* left=lowestcommonancestor(root->left,p,q);
    node* right=lowestcommonancestor(root->right,p,q);
    if(left==NULL)
    {
        return right;
    }
    else if(right==NULL)
    {
        return left;
    }
    else{
        return root;
    }
}
int main()
{
    node* root=new node(1);
    root->left=new node(2);
    root->right=new node(3);
    root->left->left=new node(4);
    root->left->right=new node(5);
    root->right->right=new node(6);
    node* p=root->left->left;
    node* q=root->left->right;
    node* ans=lowestcommonancestor(root,p,q);
    cout << ans->data;
    return 0;
}
            //         1
            //       /   \
            //     2       3
            //   /   \      \
            //  4      5      6

File: LLinJava/lowestcommonancestor.java
public class lowestcommonancestor {
    public static class node 
    {
        public int data;
        public node left;
        public node right;

        node(int data1) 
        {
            this.data = data1;
        }
    }

    public static node commonancestor(node root, node p, node q) {
        if (root == null || root == p || root == q) {
            return root;
        }

        node left = commonancestor(root.left, p, q);
        node right = commonancestor(root.right, p, q);

        if (left == null) return right;
        if (right == null) return left;
        return root;
    }

    public static void main(String[] args) {
        node root = new node(1);
        root.left = new node(2);
        root.right = new node(3);
        root.left.left = new node(4);
        root.left.right = new node(5);
        root.right.right = new node(6);

        node p = root.left.left;   // node with data 4
        node q = root.left.right;  // node with data 5

        node ans = commonancestor(root, p, q);
        System.out.println(ans.data);  // ✅ Print the data value
    }
}


